### ISSUE TITLE
App Directory Does Not Support Disabling API Body Parsing (Unlike Pages Directory)

### ISSUE THREAD

**Issue by JS-GitRepo** [vercel/next.js#54090]
### Verify canary release

- [X] I verified that the issue exists in the latest Next.js canary release

### Provide environment information

```bash
Operating System:
      Platform: win32
      Arch: x64
      Version: Windows 10 Home
    Binaries:
      Node: 18.17.1
      npm: N/A
      Yarn: N/A
      pnpm: N/A
    Relevant Packages:
      next: 13.4.12
      eslint-config-next: 13.4.12
      react: 18.2.0
      react-dom: 18.2.0
      typescript: 5.1.6
    Next.js Config:
      output: N/A
```

### Which area(s) of Next.js are affected? (leave empty if unsure)

App Router, Middleware / Edge (API routes, runtime)

### Link to the code that reproduces this issue or a replay of the bug

https://codesandbox.io/p/sandbox/keen-borg-t8lcr2?file=%2Fapp%2Fpage.tsx%3A1%2C1

### To Reproduce

1. Create Next JS project using App Directory, Route Handlers, and deploy to Firebase Hosting using Web Framework aware Firebase Tools (Latest, 
2. Try to create any request with a body to an API Route Handler; POST, PUT etc. with body contents
3. Using Next Dev or Next Start, everything will work fine
4. Publish to Firebase, suddenly get server timeout errors 
5. Try GET request; 0 issues (no body)

### Describe the Bug

Based on my research, I feel fairly confident that the bug is that Next JS is trying to parse the JSON, and then Firebase Cloud Functions is trying to parse the already parsed JSON (unsure of the exact order of who is parsing first in this scenario.) Automatic parsing cannot be turned off in Firebase (references linked below) and they **_previously_** could be disabled in the Pages/ router but this does not appear to be the case in app/ router (at least according to Next JS documentation.

The simple fix that worked in the past for this issue was being able to disable body parsing on a per file bases in the Pages directory, and I believe there may have been a fix allowing it to be done globally as well. I do not see a way to do so in App directory as the config method has been deprecated.

**Relevant Links:**
[Stack Overflow Post Outlining Issue in Detail - Pages Directory, 2019](https://stackoverflow.com/questions/63968953/why-do-i-get-a-502-gateway-error-from-nextjs-app-hosted-on-firebase-for-post-r)
https://github.com/vercel/next.js/pull/16169
[Firebase Cloud Function Request Parsing Cannot Be Disabled](https://stackoverflow.com/questions/70992856/disable-undo-firebase-cloud-functions-default-request-parsing)
[Pages Router Let you Disable Next JS Parsing - App Router, Not Documented](https://nextjs.org/docs/pages/building-your-application/routing/api-routes)
[App Directory Route Config Options - None are Body Parse](https://nextjs.org/docs/app/api-reference/file-conventions/route-segment-config)

### Expected Behavior

Post requests should at the very least be reaching the cloud function, then failing somewhere within. Instead, the request is not being denied or processed; it is causing an "unknown" server error resulting in timeout before any code is actually run.

Simply put, the post request should be reaching the function (and does fine in both next build && next dev as well as next build && next start scenarios). In a perfect world, Firebase Cloud Functions would detect and handle this automatically but they do not seem to (firebase-tools [v12.4.7](https://github.com/firebase/firebase-tools/releases/tag/v12.4.7)).

### Which browser are you using? (if relevant)

Chrome Version 115.0.5790.171 (Official Build) (64-bit)

### How are you deploying your application? (if relevant)

Firebase Deploy

<sub>[NEXT-1522](https://linear.app/vercel/issue/NEXT-1522/app-directory-does-not-support-disabling-api-body-parsing-unlike-pages)</sub>

**Comment by JS-GitRepo** [comment]
**HTTP Error Details:**
Copy as Curl:
curl "[BASE_URL_CENSORED]/signin/createaccount/api" ^
  -H "authority: [BASE_URL_CENSORED]" ^
  -H "accept: */*" ^
  -H "accept-language: en-US,en;q=0.9" ^
  -H "content-type: text/plain;charset=UTF-8" ^
  -H "dnt: 1" ^
  -H "origin: [BASE_URL_CENSORED]" ^
  -H "referer: [BASE_URL_CENSORED]/signin/createaccount" ^
  -H "sec-ch-ua: ^\^"Not/A)Brand^\^";v=^\^"99^\^", ^\^"Google Chrome^\^";v=^\^"115^\^", ^\^"Chromium^\^";v=^\^"115^\^"" ^
  -H "sec-ch-ua-mobile: ?1" ^
  -H "sec-ch-ua-platform: ^\^"Android^\^"" ^
  -H "sec-fetch-dest: empty" ^
  -H "sec-fetch-mode: cors" ^
  -H "sec-fetch-site: same-origin" ^
  -H "user-agent: Mozilla/5.0 (Linux; Android 6.0; Nexus 5 Build/MRA58N) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/115.0.0.0 Mobile Safari/537.36" ^
  --data-raw "^{^\^"password^\^":^\^"[CENSORED]",^\^"handle^\^":^\^"jake^\^",^\^"newUserAccount^\^":null^}" ^
  --compressed

**Response**:
`<!DOCTYPE html>
<html lang=en>
    <meta charset=utf-8>
    <meta name=viewport content="initial-scale=1, minimum-scale=1, width=device-width">
    <title>Error 502 (Server Error)!!1</title>
    <style>
        * {
            margin: 0;
            padding: 0
        }

        html,code {
            font: 15px/22px arial,sans-serif
        }

        html {
            background: #fff;
            color: #222;
            padding: 15px
        }

        body {
            margin: 7% auto 0;
            max-width: 390px;
            min-height: 180px;
            padding: 30px 0 15px
        }

        * > body {
            background: url(//www.google.com/images/errors/robot.png) 100% 5px no-repeat;
            padding-right: 205px
        }

        p {
            margin: 11px 0 22px;
            overflow: hidden
        }

        ins {
            color: #777;
            text-decoration: none
        }

        a img {
            border: 0
        }

        @media screen and (max-width: 772px) {
            body {
                background:none;
                margin-top: 0;
                max-width: none;
                padding-right: 0
            }
        }

        #logo {
            background: url(//www.google.com/images/branding/googlelogo/1x/googlelogo_color_150x54dp.png) no-repeat;
            margin-left: -5px
        }

        @media only screen and (min-resolution: 192dpi) {
            #logo {
                background:url(//www.google.com/images/branding/googlelogo/2x/googlelogo_color_150x54dp.png) no-repeat 0% 0%/100% 100%;
                -moz-border-image: url(//www.google.com/images/branding/googlelogo/2x/googlelogo_color_150x54dp.png) 0
            }
        }

        @media only screen and (-webkit-min-device-pixel-ratio: 2) {
            #logo {
                background:url(//www.google.com/images/branding/googlelogo/2x/googlelogo_color_150x54dp.png) no-repeat;
                -webkit-background-size: 100% 100%
            }
        }

        #logo {
            display: inline-block;
            height: 54px;
            width: 150px
        }
    </style>
    <a href=//www.google.com/>
    <span id=logo aria-label=Google></span>
</a>
<p>
    <b>502.</b>
    <ins>That’s an error.</ins>
<p>The server encountered a temporary error and could not complete your request.
<p>
    Please try again in 30 seconds.  <ins>That’s all we know.</ins>

`

**Comment by KairuDeibisu** [comment]
This is going to be really weird but try using req.text()

This is the function that returns the full body without being parsed. This is what I had to do to validate clerk web hooks using app router.

**Comment by balazsorban44** [comment]
Thanks for reporting, let me consolidate this issue into #50166 which seems to express the same issue. This way it's easier for us to handle feedback. Feel free to comment there with helpful information, or upvote :+1: the first post to show your interest in the issue. Thanks!
